projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -0,0 +1,166 @@
+        internal const string MessagingOperation = "messaging.operation";
+        internal const string MessagingSystem = "messaging.system";
+        internal const string MessagingDestination = "messaging.destination.name";
+        internal const string MessagingDestinationRoutingKey = "messaging.rabbitmq.destination.routing_key";
+        internal const string MessagingBodySize = "messaging.message.body.size";
+        internal const string MessagingEnvelopeSize = "messaging.message.envelope.size";
+        internal const string ProtocolName = "network.protocol.name";
+        internal const string ProtocolVersion = "network.protocol.version";
+        internal const string RabbitMQDeliveryTag = "messaging.rabbitmq.delivery.tag";
+
+        internal static readonly ActivitySource s_source = new ActivitySource(Name, typeof(RabbitMQActivitySource).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion);
+        public const string Name = "RabbitMQ.Client";
+        public static bool UseRoutingKeyAsOperationName { get; set; } = true;
+
+        internal static Activity Send(string routingKey, string exchange, int bodySize)
+        {
+            if (s_source.HasListeners())
+            {
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish", ActivityKind.Producer);
+                if (activity?.IsAllDataRequested == true)
+                {
+                    PopulateMessagingTags("publish", routingKey, exchange, 0, null, bodySize, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Receive(string routingKey, string exchange, ulong deliveryTag, IReadOnlyBasicProperties readOnlyBasicProperties, int bodySize)
+        {
+            if (s_source.HasListeners())
+            {
+                // Extract the PropagationContext of the upstream parent from the message headers.
+                DistributedContextPropagator.Current.ExtractTraceIdAndState(readOnlyBasicProperties, ExtractTraceIdAndState, out string traceparent, out string traceState);
+                IEnumerable<KeyValuePair<string, string>> baggage = DistributedContextPropagator.Current.ExtractBaggage(readOnlyBasicProperties, ExtractTraceIdAndState);
+                ActivityContext.TryParse(traceparent, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{routingKey} receive" : "receive", ActivityKind.Consumer, parentContext);
+                if (activity != null && activity.IsAllDataRequested)
+                {
+                    if (baggage != null)
+                    {
+                        foreach (var item in baggage)
+                        {
+                            activity.SetBaggage(item.Key, item.Value);
```

Comment:
just FYI - this is not the Baggage API recommended by OpenTelemetry .NET.

Reply:
I'll take a closer look

Reply:
I'm a bit confused here. I need to read up on it a little bit more. I'll remove the Baggage handling in the meantime so we can add it correctly later on.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Api#baggage-api Some reference when you come back to this.